sno/pk_generation.py

```diff
@@ -144,25 +184,138 @@ def _init_schema(self):
+                    buffered_insert_limit = 0
+
+            else:
+                # New feature. Assign it a new PK and yield it.
+                yield self._assign_pk(feature, next_new_pk, feature_hash=feature_hash)
+                next_new_pk += 1
+
+        if buffered_inserts:
+
+            # Look for matching inserts-deletes - reassign the PK from the delete, treat is as an edit:
+            yield from self._match_similar_features(
+                self._find_deleted_features(hash_to_unassigned_pks), buffered_inserts
+            )
+            # Just assign new PKs to those we couldn't find a match for.
+            yield from self._assign_pk_range(buffered_inserts, next_new_pk)
+
+    def _assign_pk_range(self, features, pk):
+        for feature in features:
+            yield self._assign_pk(feature, pk)
+            pk += 1
+
+    def _assign_pk(self, feature, pk, feature_hash=None):
+        if feature_hash is None:
+            feature_hash = self.schema.hash_feature(feature, without_pk=True)
+
+        feature[self.primary_key] = pk
         self.pk_to_hash[pk] = feature_hash
-        return pk
+        return feature
+
+    def _match_similar_features(self, old_features, new_features):
+        orig_old_features_len = len(old_features)
+        orig_new_features_len = len(new_features)
+        similar_count = 0
+
+        for old_feature, new_feature in self._extract_similar_pairs(
+            old_features, new_features
+        ):
+            pk = old_feature[self.primary_key]
+            yield self._assign_pk(new_feature, pk)
+            similar_count += 1
+
+        assert len(old_features) == orig_old_features_len - similar_count
+        assert len(new_features) == orig_new_features_len - similar_count
+
+    def _extract_similar_pairs(self, old_features, new_features):
+        # Copy old_features so we can remove from it while iterating over it:
+        for old_feature in old_features.copy():
+            new_feature = self._find_sole_similar(old_feature, new_features)
+            if (
+                new_feature is not None
+                and self._find_sole_similar(new_feature, old_features) is not None
+            ):
+                old_features.remove(old_feature)
+                new_features.remove(new_feature)
+                yield old_feature, new_feature
+
+    def _find_sole_similar(self, target, source_list):
+        match_count = 0
+        for s in source_list:
+            if self._is_similar(s, target):
+                match = s
+                match_count += 1
+                if match_count > 1:
+                    break
+
+        return match if match_count == 1 else None
+
+    def _is_similar(self, lhs, rhs):
+        # NOTE: This is one of several possible similarity metrics.
+        # TODO: Add more and make them configurable, if this proves useful.
+
+        dissimilar_count = 0
+
+        for l, r in zip_longest(lhs.values(), rhs.values()):
```

Comment:
are these dicts _guaranteed_ to be in the same order? what guarantees that?

Reply:
I personally guarantee it. You have my word.
I just added this guarantee to the two main interfaces that define features in Sno:
`base_dataset.py` and `import_source.py`